### PR TITLE
Infinite Scroll test number of items generated

### PR DESCRIPTION
**Pull request by jcfilben**
#### What does this PR do?
Adds testing to infinite scroll `step` prop. Tests small, medium, and large step sizes. Ensures that the number of elements being rendered is correct.

NOTE: this also tests that only unique items are being rendered, (no duplicates). On the `length when step < array` test if you uncomment `show={117}` the test will generate a warning and fail because of a bug causing there to be duplicate items when `show` is set and it is a higher value than `step`.

#### How should this be manually tested?
`yarn test` before and after uncommenting line 89

#### Screenshots (if appropriate)
Screen Shot 2020-08-07 at 4 49 44 PM

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
No

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

### CODE REVIEW

**Comment by ShimiSun** [comment]
Relevant issues #4369 , #4334

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js [hunk 1]
```diff
@@ -75,4 +76,52 @@ describe('InfiniteScroll', () => {
...
+          step={step}
+        >
+          {item => <Box key={item}>{item}</Box>}
+        </InfiniteScroll>
+      </Grommet>,
+    );
+    expect(container.firstChild.children.length).toEqual(step + 1);
+  });
+
+  test('length when step = array', () => {
```
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:99]: Nice. Was going to suggest adding this case.

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js [hunk 2]
```diff
@@ -75,4 +76,52 @@ describe('InfiniteScroll', () => {
     );
     expect(container.firstChild).toMatchSnapshot();
   });
+
+  test('length when step < array', () => {
+    const allItems = Array(1000)
+      .fill()
+      .map((_, i) => `item ${i + 1}`);
```
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:83]: Since this seems like we are repeating this logic across several tests, does it make sense for us to abstract this and call it something like `simpleItems`?

Or is the direction to keep the test completely self contained?

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js [hunk 3]
```diff
@@ -75,4 +76,52 @@ describe('InfiniteScroll', () => {
     );
     expect(container.firstChild).toMatchSnapshot();
   });
+
+  test('length when step < array', () => {
```
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:80]: Reading the descriptions/names of the tests, I am able to understand the test scenario, but it is not immediately clear what the expected results or outcome should be.

Could these be written in the form of "Should [expectation / outcome] when [scenario]?" For example, "Should render items equal to the length of step when step < items.length" or something similar. I can reason with the code to try to interpret the intent, but naming in this form might be helpful. Thoughts?

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js [hunk 4]
```diff
@@ -75,4 +76,52 @@ describe('InfiniteScroll', () => {
...
+        <InfiniteScroll
+          items={allItems}
+          // show={117}
+          step={step}
+        >
+          {item => <Box key={item}>{item}</Box>}
+        </InfiniteScroll>
+      </Grommet>,
+    );
+    expect(container.firstChild.children.length).toEqual(step + 1);
```
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:96]: Why `step + 1`? If I am unfamiliar with how this component is implemented, then I may not have context for why adding 1 to the step. A comment could be helpful here.

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js (hunk 3 above)
Comment (IanKBovard) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:80]: I agree, I think the `should ---> when` format provides a good amount of description.

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js (hunk 4 above)
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:96]: This is a style suggestion which I think helps readability and intent for those who follow afterward. It's a bit more verbose, but helps communicate what each of these values represent.

```suggestion
    const renderedItems = container.firstChild.children.length;
    const expectedItems = step + 1 // comment explaining why + 1
    expect(renderedItems).toEqual(expectedItems);
```

Comment (IanKBovard) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:96]: i could be wrong, but i believe is because it'll render items 0(inclusive)-50(inclusive)? making it 51 items rendered.

Comment (halocline) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:96]: I think in this case, the 51st element is actually a 0 height ref to mark the end of the page of results.

Comment (IanKBovard) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:96]: ah, yes. that makes more sense

Comment (jcfilben) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:96]: I added comments about `step + 1`. Let me know if they are too long or could be phrased better

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js (hunk 2 above)
Comment (jcfilben) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: Yes I agree it makes sense to get rid of the repeating logic. I added `simpleItems` at the beginning of the test suite so it can be used by all 3 of the tests instead of being duplicated

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js (hunk 3 above)
Comment (jcfilben) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: I renamed the tests to variations of `Should render items equal to the length of 
  step + 1 when step < items.length`. I agree it is more descriptive than what the tests were named previously.

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js (hunk 4 above)
Comment (jcfilben) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: Updated tests to include this style suggestion

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js [hunk 5]
```diff
@@ -75,4 +80,62 @@ describe('InfiniteScroll', () => {
...
+          {item => <Box key={item}>{item}</Box>}
+        </InfiniteScroll>
+      </Grommet>,
+    );
+    // When step > the length of the number of items,
+    // an additional item is NOT rendered because the items
+    // don't approach the end of the page. This is why the total
+    // number of items is the length of simpleItems, 1000.
+    const renderedItems = container.firstChild.children.length;
+    const expectedItems = 1000;
```
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:138]: What would happen if simpleItems was redefined at some point in the future?

```suggestion
    const expectedItems = simpleItems.length;
```

Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:138]: Rather than hardcoding 1000 twice, can we `const numItems = 1000;` and then have `simpleItems(numItems)` and expectedItems = numItems?

Comment (jcfilben) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: Yes, fixed this in latest push

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js [hunk 6]
```diff
@@ -75,4 +80,62 @@ describe('InfiniteScroll', () => {
     );
     expect(container.firstChild).toMatchSnapshot();
   });
+
+  test(`Should render items equal to the length of 
+  step + 1 when step < items.length`, () => {
+    const step = 50;
```
Comment (halocline) [inline src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js:86]: Just calling this out to set stage for future context:

As our test scenarios get more complex, we will likely see a couple of other marker refs being included in the array. We will also likely want to find a way to re-use and reference these three tests by other tests.

As we write those scenarios, these tests may get refactored to accommodate. That said, I think the current state helps us move forward and we can refactor as needed.

File: src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js (hunk 4 above)
Comment (halocline) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: I think what we have here is good for now. We may revisit as we introduce additional tests and try to repurpose what is being done here.

For context, what '1' is representing here is an additional ref to serve as an anchor/trigger for additional functionality. As soon as we add a prop such as `show` there will be an additional ref marking the show index position in the middle of the resulting items and so the children in the array would then be step + 2.

@IanKBovard can you think of a way where we can test for `renderedItems.length === step`? Perhaps we do something like:
```
    const allChildren = container.firstChild.children;
    // Removing any children which are serving as refs
    const refKeys = ["first", "last", "show"];
    const pageItems = allChildren.filter( /* some sort of filter logic which returns children where there key is not in refKeys */ )
    const expectedItems = step;
    expect(pageItems.length).toEqual(expectedItems);
```

Comment (IanKBovard) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: I agree. Filter is probably the best approach.

Comment (halocline) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: @jcfilben Can you take a look at the suggested approach above. If it makes sense, would you mind taking a stab at implementing?

Comment (IanKBovard) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: If possible, i think we can move all creating and filtering of children logic in a before block when we start to refactor these in future prs.

Comment (jcfilben) [reply src/js/components/InfiniteScroll/__tests__/InfiniteScroll-test.js]: Yup, just updated the PR using the approach outlined above
